piff/optical_model.py

```diff
@@ -175,52 +268,211 @@ def fit(self, star):
-            gaussian = galsim.Gaussian(sigma=self.sigma)
-            prof.append(gaussian)
-
-        # atmosphere
-        if len(self.atm_kwargs) > 0:
-            if 'L0' in self.atm_kwargs and self.atm_kwargs['L0'] is not None:
-                atm = galsim.VonKarman(**self.atm_kwargs)
-            else:
-                atm = galsim.Kolmogorov(**self.atm_kwargs)
-            prof.append(atm)
 
-        # optics
-        if params is None or len(params) == 0:
-            # no aberrations.  Just the basic opt_kwargs
-            optics = galsim.OpticalPSF(**self.opt_kwargs)
+        if self.mirror_figure_screen:
+
+            optical_screen = galsim.OpticalScreen(diam=self.diam, aberrations=zernike_coeff, lam_0=self.lam)
+            screens = galsim.PhaseScreenList([optical_screen, self.mirror_figure_screen])
+            optics_psf = galsim.PhaseScreenPSF(screen_list=screens, aper=self.aperture, lam=self.lam, gsparams=self.gsparams
+            )
+
         else:
-            aberrations = [0,0,0,0] + list(params)
-            optics = galsim.OpticalPSF(aberrations=aberrations, **self.opt_kwargs)
+            optics_psf = galsim.OpticalPSF(
+                lam=self.lam, diam=self.diam, aper=self.aperture,
+                aberrations=zernike_coeff, gsparams=self.gsparams
+            )
+
+        return optics_psf
 
-        # convolve together
-        prof.append(optics)
+    @lru_cache(maxsize=8)
+    def getAtmosphere(self, r0, L0=None, g1=0.0, g2=0.0):
+        """ Get the Atmospheric PSF component.
 
-        if len(prof) == 1:
-            prof = prof[0]
+        :param r0:          Fried parameter
+        :param L0:          Outer scale
+        :param g1:          Shear g1 component
+        :param g2:          Shear g2 component
+
+        :returns: a galsim.GSObject instance
+        """
+        if self.atmo_type == 'VonKarman':
+            if L0 is None or L0==0.:
+                raise ValueError("No value specified for VonKarman L0 ")
+            else:
+                atm = galsim.VonKarman(lam=self.lam, r0=r0, L0=L0, flux=1.0, gsparams=self.gsparams)
+        elif self.atmo_type == 'Kolmogorov':
+            atm = galsim.Kolmogorov(lam=self.lam, r0=r0, flux=1.0, gsparams=self.gsparams)
+        elif self.atmo_type == 'None' or self.atmo_type == None:
+            atm = galsim.DeltaFunction()
         else:
-            prof = galsim.Convolve(prof)
+            raise ValueError("Invalid atmo_type ",self.atmo_type)
+
+        # shear
+        if g1!=0.0 or g2!=0.0:
+            atm = atm.shear(g1=g1, g2=g2)
+        return atm
+
+    def unpack_params(self,params):
+        """Unpack parameter values
+
+        :param params               An ndarray or list with parameters in order
+
+        :returns: zernike_coeff,r0,L0,g1,g2
+        """
+        kwargs = self.params_tokwargs(params)
+        return kwargs['zernike_coeff'],kwargs['r0'],kwargs['L0'],kwargs['g1'],kwargs['g2']
+
+    def params_tokwargs(self,params):
```

Comment:
More standard name would be `params_to_kwargs`.  Likewise for `kwargs_to_params`.